---
title: "AMI analysis: task123"
author: "Q.Zhang"
date: "10/14/2017"
output: 
  html_document: 
    theme: united
    highlight: tango
    toc: true
    toc_float: true
    number_sections: true
---
Here we present a standard pipeline for analysing microarray data. Raw microarry data was deposited in NCBI Gene Expression Omnibus (GEO) with accession number [GSE48060](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48060). There were total of 31 acute myocardial infarction samples(AMI) with 21 normal controls. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = TRUE, warning = FALSE, include = TRUE, tidy = TRUE, message = FALSE, cache = TRUE)
setwd('~/b_affy')
```

<br>

# Get data and labels

```{r getdata,  eval=FALSE}
## ------ get data from GEO -------
library(GEOquery)
getGEOSuppFiles("GSE48060")
untar("GSE48060/GSE48060_RAW.tar", exdir="data")
cels <- list.files("data/", pattern = "[gz]")
sapply(paste("data", cels, sep="/"), gunzip)

```

To make label for each sample, GDS5074 soft files was downloaded into local directory. `ph` object was made for labeling those samples. Here is a brief sumary of the information (eg. platform, sample labels) from related `.soft` file.

```{r labels}
## ------ make labels --------
# GDS soft files are manually downloaded from GEO
library(dplyr)
library(GEOquery)

gds5074 <- getGEO(filename='~/b_affy/GDS5074.soft.gz')
eset <- GDS2eSet(gds5074)
pheno_raw <- pData(eset) %>% tbl_df()
# make phenodata object
files <- list.files(path = 'data')

ph <- tibble(sample_name = pheno_raw$sample,
             file_name = files[mapply(function(p,x) grep(p,x),
                                      p = pheno_raw$sample,MoreArgs = list(x = files))],
             disease_state = c(rep('AMI', 31), rep("control", 21)),
             recur = NA) %>% dplyr::select(file_name, sample_name, disease_state, recur)

# modify the labels
mod.labels <- function(dt){
    dt$recur[grepl('without', pheno_raw$description)] <- 'AMI_no'
    dt$recur[grepl('control', pheno_raw$description)] <- 'control'        
    dt$recur[is.na(dt$recur)] <- 'AMI_yes'
    dt
}
ph <- mod.labels(ph)

```

So here we present a brief summary of what is in the dataset! For simplicity here we present only the top 6 sample labels in `ph`.

```{r printsth }
pander::pander(Meta(gds5074))
knitr::kable(head(ph))
```


<br>



# gcrma normalize

GC Robust Multi-array Average (GCRMA) is a method of converting `.CEL` files into expression set using the Robust Multi-array Average (RMA) with the help of probe sequence and with GC-content background correction. Although the original paper used RMA, we confirmed that GCRMA gives more contrast in fold change in differential expression between AMI and control samples, which maybe more preferable for making bayesian network classifiers.

Here `data.matrix` gives the input for DE analysis.

```{r gcrma}
library(simpleaffy)
library(genefilter)
celfiles <- read.affy(covdesc="phenodata.txt", path="data")     # affybatch object

gcnorm_filter <- function(cf){
  cf.gcrma <- gcrma(cf)
  cf.filtered <- nsFilter(cf.gcrma)
  print(cf.filtered$filter.log)
  return(cf.filtered)
}

cf.gf <- gcnorm_filter(cf = celfiles)
data.matrix <- cf.gf$eset %>% exprs

```

<br>


# clustering

Since we are facing a copious of probes, dimension reduction becomes inevitable. Here we present both PCA and t-sne method to reduce to a feasible dimmension size, and use the reduced features to further cluster AMI samples and controls.

First PCA is used, however the combination of first three dimensions did not give satisfactory divergence between AMI and controls. Hence the resulting dendrogram did not separate two classes well enough (see the graph below).

Then we turn to tsne, which is aims for dimensionality reduction that is particularly well suited for the visualization of high-dimensional datasets. However such divergence is questionable if we permuate seed, which suggest tsne is not yet robust. This result somehow suggested the need for feature selection before constructing bayesian network classifier.

In the following plots, red represents AMI samples while green refers to the controls.

```{r dr_cluster}
library(Rtsne)
library(magrittr)

make.pca <- function(dm = data.martix){
  data.PC = prcomp(t(data.matrix),scale.=FALSE)
  PCdf <- data.PC$x %>%  as.data.frame() 
  PCdf
}

PCdf <- make.pca()
color=c(rep('red', 31), rep('green', 21))
pairs(~PC1 + PC2 + PC3,data = PCdf, 
      main="PCA", col = color) 


labels <- c(rep('AMI', 31), rep('control', 21))
set.seed(3)
make.tsne <- function(df = t(data.matrix), lb = labels){
  data.tsne <- Rtsne(df, dims = 2, perplexity=7, verbose=TRUE, max_iter = 500)
  tsne_df <- as.data.frame(data.tsne$Y)
  tsne_df
}


tsne.plot <- function(lb = labels, dt = tsne_df){
  plot(dt, t='n', main="tsne")
  text(dt, labels=lb, col=color, cex = 0.8)
}

tsne_df <- make.tsne()
tsne.plot()

```

Despite the distinction in 2D plot is not satisfactory, we here give hclust to illustrate how such dimension reduction might help with proper clustering of datasets.

Here both methods give postive result, however t_SNE give better clustered chunks than PCA. This suggests this dataset is worth of further learning.

```{r hclust}
library(dendextend)
fseq <- colnames(data.matrix)

mod.df <- function(raw_df, fs = fseq, phdict = ph){
  rownames(raw_df) <- phdict$sample_name[match(fs, phdict$file_name)]
  raw_df
}

hclust_pipe <- function(df, method){
  distance <- dist(df, method = 'maximum')
  dendo <- hclust(distance) %>% as.dendrogram()
  lb <- labels(dendo)
  col <- ifelse(ph$disease_state[match(lb, ph$sample_name)]=='AMI', 'red', 'green')
  dendo %>% set('labels', lb) %>% 
  set("labels_col", col) %>% 
  set("labels_cex", 1) %>% 
  plot(main = paste("hierarchical clustering by",method) )
  
}

PCdf <- mod.df(PCdf)
tsne_df <- mod.df(tsne_df)
hclust_pipe(PCdf,method = 'PCA')
hclust_pipe(tsne_df, method = 't-SNE')
```

## future work for this section: 
We will apply `caret` package for feature selection prior to any dimensio reduction techniques, but now there is some installation problem here...


<br>


# limma for differential expression analysis
Here we make design matrix for contrasting AMI from control samples and selected the differentially expressed genes by lfc cutoff 0.5(1.4 fold change). This table is further annotated with many gene identifiers for subsequent analysis of pathway and ontologies.

For better visualization here only a subset of differentially expressed genes are shown.

```{r limma}

library(limma)
library(hgu133plus2.db)
library(annotate)

DE_pipe <- function(cff = cf.gf){
  # make matrix
  samples <- ph$disease_state %>% as.factor()
  design <- model.matrix(~0 + samples)
  colnames(design) <- c("AMI", "control")
  fit <- lmFit(exprs(cff$eset), design)
  contrast.matrix <- makeContrasts(AMI-control, levels=design)
  # compare
  huvec_fits <- contrasts.fit(fit, contrast.matrix)
  huvec_ebFit <- eBayes(huvec_fits)
  tab <- topTable(huvec_ebFit, coef=1,lfc = 0.5, number = 1000)   
}

anno <- function(chip = hgu133plus2.db, t = tab){
  gs2 <- AnnotationDbi::select(x = chip, keys = rownames(tab), keytype = 'PROBEID',
                             columns = c('ENTREZID','GENENAME','SYMBOL')) %>% tbl_df()
  tab$PROBEID <- rownames(tab)
  tab <- left_join(tab, gs2) %>% tbl_df()
}

tab <- DE_pipe()
tab <- anno()


tab <- tab %>% dplyr::select(logFC, P.Value, ENTREZID, GENENAME, SYMBOL, PROBEID )
knitr::kable(head(tab))
```

<br>


# bicluster heatmap
For better visualization, here we only present differentially expressed genes with p.val < 0.0005 as proof of concept. The idea of bicluster heatmap is a side-proof of differential expressed genes, where the ones who are upregulated in population A, should be downregulated in population B.

```{r heat}
library(d3heatmap)
library(preprocessCore)

normf <- function(m = data.matrix, de = tab){
  tab0 <- de[ order(-tab$logFC), ] 
  tab0 <- tab0[tab0$P.Value < 0.0005 & abs(tab0$logFC) > 0.6,]
  de_pids <- tab0$PROBEID
  dem <- m[de_pids,]
  colnames_dem <- ph$sample_name[match(colnames(dem), ph$file_name)]
  rownames_dem <- de$ENTREZID[match(rownames(dem), de$PROBEID)]
  # normalize in 2D
  dem <- t(normalize.quantiles(t(dem)))
  dem <- normalize.quantiles(dem)
  dimnames(dem) <- list(rownames_dem, colnames_dem)
  dem
}


dem <- normf()
d3heatmap(dem, scale = "none", dendrogram = "none" )
```

# Pathways and ontologies
The above differential expression table is then imported into [Reactome](reactome.org) browser, and here is the briefing of the result with pValue < 0.05 and FDR < 0.5. Here we use gprofilerR API for GO enrichment. 

The original paper used GenGo Metacore software (which requires purchasing) and identified siz pathway related to `immune response`, featuring `apoptosis`, `coagulation` and `interleukin signaling`, which agrees quite well with our result despite minor divergence.

```{r reactome }
library(readr)
ptw <- read_csv('bn_mod/reactome_result.csv')
ptw1 <- ptw %>% dplyr::select(`Pathway identifier`, `Pathway name`, `Entities pValue`, `Entities FDR`) %>% dplyr::filter(`Entities pValue` < 0.05 & `Entities FDR` < 0.5)

knitr::kable(ptw1)
```

For simplicity only the top six most significantly enriched GO terms were displayed. Those GO terms are somehow more biologically meaningful than the ones on the original paper, which are big terms including `Positive regulation of biological process`, `Cellular metabolic process`, `Primary metabolic process`, `Immune system process` and `Metabolic process`. Meanwhile our enriched GO terms feature more on `immunology`, which is consistent with pathway analysis from reactome.

```{r gprofiler}
library(gProfileR)
library(dplyr)
genes <- tab$SYMBOL
gpgenes <- gprofiler(genes) %>% tbl_df()

gp <- gpgenes %>% dplyr::select(term.id, term.name, domain, p.value) %>% dplyr::filter(p.value < 0.01) 
gp1 <- gp[match(sort(gp$p.value),gp$p.value),]
gp1$p.value <- as.character(gp1$p.value)# for sake of visualization
knitr::kable(head(gp1))

```

# future work
Now we are trying different method for feature discretization and network consturction. Many thanks to Yufeng Zhang, Haipeng Wang for their help and encouragement, it has been one of the happiest days working with them. Thanks Gil for offering me this oppotunity. For the latest progress update please check [this repo](https://github.com/hurrialice/b_affy).
```{r end}
sessionInfo()
```